File: blind 75 leetcode/Arrays and Strings/two_sum.py
'''
Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

You can return the answer in any order.

 

Example 1:

Input: nums = [2,7,11,15], target = 9
Output: [0,1]
Output: Because nums[0] + nums[1] == 9, we return [0, 1].
Example 2:

Input: nums = [3,2,4], target = 6
Output: [1,2]
Example 3:

Input: nums = [3,3], target = 6
Output: [0,1]
'''

def twoSum(self, nums: List[int], target : int) -> List[int]:
	## map the target value - the current number with index of the current number
	pairs = dict()
	for i in range(len(nums)):
		## if the current number is found in the pairs dictionary...
		if (nums[i] in pairs):
			## then a pair has been found
			## return the indices of the two numbers
			return [pairs[nums[i]], i]
		else:
			## store the index of the number, and its potential complementary pair
			## by subtracting the target value with the current number
			pairs[target - nums[i]] = i 
	return None

File: neetcode java implementations/twoSum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
        int[] result = new int[2];
        for (int i = 0; i < nums.length; ++i){
            int other = target - nums[i];
            if (map.get(other) != null){
                int index = map.get(other);
                if (index < i){
                    result[0] = index;
                    result[1] = i;
                }
                else {
                    result[0] = i;
                    result[1] = index;
                }
                return result;
            }
            else {
                map.put(nums[i], i);
            }
        }
        return result;
    }
}
